rotordc_sale_picking_state/models/sale_order.py

```diff
@@ -0,0 +1,74 @@
+# SPDX-FileCopyrightText: 2022 Coop IT Easy SC
+#
+# SPDX-License-Identifier: AGPL-3.0-or-later
+
+from odoo import api, fields, models
+from odoo.tools.translate import _
+
+
+class SaleOrder(models.Model):
+    _inherit = "sale.order"
+
+    @api.model
+    def get_picking_state_selection(self):
+        return [
+            # No stock picking found.
+            ("none", _("None")),
+            # Multiple conflicting found.
+            ("unknown", _("Unknown")),
+            # Same as states on stock.picking.
+            # FIXME: Should we dynamically create these from stock.picking? What
+            # about translations?
+            ("draft", _("Draft")),
+            ("waiting", _("Waiting Another Operation")),
+            ("confirmed", _("Waiting")),
+            ("assigned", _("Ready")),
+            ("done", _("Done")),
+            ("cancel", _("Cancelled")),
+        ]
+
+    @api.depends(
+        "picking_ids.picking_type_code",
+        "picking_ids.state",
+    )
+    def _compute_picking_states(self):
+        for order in self:
+            # Map picking_type_codes to states.
+            code_mapping = {}
+            conflicts = set()
+            for picking in order.picking_ids:
+                # The picking's state is either the first detected state for a
+                # given code, or the same state has been encountered before for
+                # that code.
+                if (
+                    code_mapping.setdefault(picking.picking_type_code, picking.state)
+                    == picking.state
+                ):
+                    continue
+                # picking_type_code has conflicting states.
+                else:
+                    conflicts.add(picking.picking_type_code)
```

Comment:
It might be a matter of taste, but I prefer more simpler conditions like
```suggestion
if picking.picking_type_code not in code_mapping:
    code_mapping[picking.picking_type_code] = picking.state
if code_mapping[picking.picking_type_code] != picking.state:
    conflicts.add(picking.picking_type_code)
```
but it's very subjective ^^

Reply:
You're probably right that your suggestion is more legible. `code_mapping.setdefault(picking.picking_type_code, picking.state) == picking.state` does a few too many things in one line of code. It hurts, because I love `setdefault` :sweat_smile:

Reply:
done